modules/system/files/linker.py

```diff
@@ -0,0 +1,185 @@
+        continue
+    if path.exists(follow_symlinks=False):
+        # There is a file at this path
+        # It could be a regular file or a symlink (including broken symlinks)
+
+        if path.is_symlink():
+            # The file is a symlink
+
+            if old_file.type != "link":
+                # This files wasn't a link at last activation, which means that the user changed it
+                # Therefore we don't touch it
+                continue
+
+            # Check that its destination remains the same
+            link_target = path.readlink()
+            if path.parent / link_target == old_file.source:
+                # The link has not changed since last system activation, so we can overwrite it
+                transactions.append(Transaction(path, path, "remove"))
+            else:
+                # The link is to somewhere else, so leave it alone
+                continue
+        else:
+            # The file is a regular file
+            continue
+    else:
+        # There's no file at this path anymore, so we have nothing to do anyway
+        continue
+
+enclosingDirectories: list[Path] = []
+# Perform all transactions
+for t in transactions:
+    # NOTE: the naming scheme for transaction properties is confusing
+    # We are **NOT** using the same scheme as symlinks when we talk about
+    # source/destination. The way we are using these names, `source` is a path
+    # in the Nix store, and `destination` is the path in the system where the source
+    # should be linked or copied to.
+    # In the special case of removing files, `destination` can be ignored
+    if DRY_RUN:
+        match t.type:
+            case "link":
+                print(f"ln -s {t.source} {t.destination}")
+            case "remove":
+                print(f"rm {t.source}")
+            case _:
+                print(f"Unknown transaction type {t.type}")
+    else:
+        match t.type:
+            case "link":
+                # Ensure parent directory exists
+                path.parent.mkdir(parents=True,exist_ok=True)
+                # Remove the file if it exists (we should only get to this case if it's an old symlink we're replacing)
+                # This does not properly handle race conditions, but I think we'd fail in the checking stage
+                #   if your config has a race condition
+                t.destination.unlink(missing_ok=True)
+                # Link the file into place
+                t.destination.symlink_to(t.source)
+            case "remove":
+                enclosingDirectories.append(t.source.parent)
+                t.source.unlink()
+            case _:
+                print(f"Unknown transaction type {t.type}")
+
+for directory in enclosingDirectories:
+    if not directory.is_dir():
+        continue
+    if any(directory.iterdir()):
+        continue
+    print(f"The directory {directory} has been emptied; you may want to remove it")
```

Comment:
Maybe this should be printed in color or otherwise made more attention-grabbing?